pkg/cli/term/reader_windows.go

```diff
@@ -106,13 +107,67 @@ var keyCodeToRune = map[uint16]rune{
 // A subset of constants listed in
 // https://docs.microsoft.com/en-us/windows/console/key-event-record-str
 const (
-	leftAlt   = 0x02
-	leftCtrl  = 0x08
-	rightAlt  = 0x01
-	rightCtrl = 0x04
-	shift     = 0x10
+	leftAlt   uint32 = 0x02
+	leftCtrl  uint32 = 0x08
+	rightAlt  uint32 = 0x01
+	rightCtrl uint32 = 0x04
+	shift     uint32 = 0x10
 )
 
+type HKL (uint32)
+
+type keyboardLayout struct {
+	lastCheck time.Time
+	layout    HKL
+	hasAltGr  bool
+}
+
+var (
+	user32                                = syscall.NewLazyDLL("user32.dll")
+	procGetKeyboardLayout                 = user32.NewProc("GetKeyboardLayout")
+	procVkKeyScanExA                      = user32.NewProc("VkKeyScanExA")
+	currentKeyboardLayout *keyboardLayout = nil
+)
+
+const keyboadLayoutCheckTimeout = time.Duration(1e+9) // 1 second
```

Comment:
Thanks for the comment but I'm still unhappy with the need for a seemingly magic constant (the `1e+9` component). Not to mention that doing the check roughly once a second seems like it is arbitrary and either not quite often enough or too often depending on whether you think people frequently change their keyboard definition or rarely do so. Ideally this would be done via a OS callback that runs whenever the user changes the keyboard definition. For example, does MS Windows provide a Go compatible signal that could be used to change the Elvish keyboard layout only when the user does change the layout as an alternative to polling? Is polling the accepted way to handle this in MS Windows apps?

Reply:
> [...] seems like it is arbitrary [...]

It is. Changing the keyboard layout usually isn't done frequently (probably mostly for compatibility reasons with older software). I chose one second to not have too much delay between changing the layout and Elvish noticing the change, but also to not check too frequently.

> [...] does MS Windows provide a Go compatible signal [...]

Windows only provides `WM_INPUTLANGCHANGE`, but since it's a message for the UI subsystem, I'd have to handle another main loop. Do you have a suggestion of which package to use?

Reply:
> Do you have a suggestion of which package to use?

Sorry, no. I haven't done any substantive Windows programming in twenty years.

Do Windows users really change their keyboard layout with any frequency other than configuring it once? Would it make more sense to detect the layout just once in setup_windows.setup() rather than doing it every four or five chars typed by the user (assuming normal typing speeds)?

Reply:
Also, instead of `time.Duration(1e+9)` just use the constant `time.Second`. Assuming you retain this var rather than switch to doing the keyboard layout setup just once.

Reply:
I personally only change keyboard layouts on accident by pressing `Alt + Shift`. However, I think that this feature has been implemented for a reason.

Then again, the only thing the code does with an AltGr layout is allowing runes to be generated from `Ctrl + Alt + <Key>`, which probably doesn't happen on non-AltGr layouts.

TL;DR: Yes, I think it would suffice to simply detect the layout once on startup. I'll make the change.